src/controller/ai_controller.c

```diff
@@ -93,9 +108,176 @@ int char_to_act(int ch, int direction) {
     return ACT_STOP;
 }
 
+/** 
+ * \brief Roll chance for pilot preference.
+ *
+ * \param pref_val The value of the pilot preference (-400 to 400)
+ *
+ * \return A boolean indicating whether the preference is confirmed.
+ */
+bool roll_pref(int pref_val) {
+    int rand_roll = rand_int(800);
+    int pref_thresh = pref_val + 400;
+    return rand_roll <= pref_thresh;
+}
+
+/** 
+ * \brief Determine whether the AI is smart enough to usually go ahead with an action.
+ *
+ * \param a The AI instance.
+ *
+ * \return A boolean indicating whether the AI is smart enough.
+ */
+bool smart_usually(ai *a) {
+    int rand_roll = rand_int(32);
+    int diff_factor = a->difficulty * a->difficulty; // 1 - 49
+    return rand_roll <= diff_factor;
+}
+
+/** 
+ * \brief Determine whether the AI is dumb enough to usually go ahead with an action.
+ *
+ * \param a The AI instance.
+ *
+ * \return A boolean indicating whether the AI is dumb enough.
+ */
+bool dumb_usually(ai *a) {
+    return !smart_usually(a);
+}
+
+/** 
+ * \brief Determine whether the AI is smart enough to sometimes go ahead with an action.
+ *
+ * \param a The AI instance.
+ *
+ * \return A boolean indicating whether the AI is smart enough.
+ */
+bool smart_sometimes(ai *a) {
+    int rand_roll = rand_int(72);
+    int diff_factor = a->difficulty * a->difficulty; // 1-49
+    return rand_roll <= diff_factor;
+}
+
+/** 
+ * \brief Determine whether the AI is dumb enough to sometimes go ahead with an action.
+ *
+ * \param a The AI instance.
+ *
+ * \return A boolean indicating whether the AI is dumb enough.
+ */
+bool dumb_sometimes(ai *a) {
+    return !smart_sometimes(a);
+}
+
+int is_special_move(af_move *move) {
+    const char *move_string = str_c(&move->move_string);
```

Comment:
Instead of converting the "str" to C string, it might be worth it to add a helper function to string handling library (str.c/.h), eg. "str_equal_c" or somesuch, which matches str against C string. Or perhaps a function to matching against multiple strings ? This pattern of needing to match lots of strings seems to repeat in the codebase, so it might be of some value later.

This is not a requirement, just a thought. Up to you.

Reply:
I just went with str_equal_c for now.

Commits:

https://github.com/omf2097/openomf/pull/387/commits/88bf16be4dc2503b64c508b8d44bee2b6967eb1a
https://github.com/omf2097/openomf/pull/387/commits/d421ca7a72567517a08f4d4750967a8841f63f4d